operators/endpointmetrics/pkg/util/status.go

```diff
@@ -10,45 +10,92 @@ import (
 )
 
+type StatusConditionName string
+
+const (
+	DeployedStatus           StatusConditionName = "Deployed"
+	DisabledStatus           StatusConditionName = "Disabled"
+	DegradedStatus           StatusConditionName = "Degraded"
+	NotSupportedStatus       StatusConditionName = "NotSupported"
+	MaxStatusConditionsCount                     = 10
+)
+
 var (
-	conditions = map[string]map[string]string{
-		"Deployed": {
-			"type":    "Progressing",
-			"reason":  "Deployed",
-			"message": "Metrics collector deployed"},
-		"Disabled": {
-			"type":    "Disabled",
-			"reason":  "Disabled",
-			"message": "enableMetrics is set to False"},
-		"Degraded": {
-			"type":    "Degraded",
-			"reason":  "Degraded",
-			"message": "Metrics collector deployment not successful"},
-		"NotSupported": {
-			"type":    "NotSupported",
-			"reason":  "NotSupported",
-			"message": "No Prometheus service found in this cluster"},
+	conditions = map[StatusConditionName]*oav1beta1.StatusCondition{
+		DeployedStatus: {
+			Type:    "Progressing",
+			Reason:  "Deployed",
+			Message: "Metrics collector deployed",
+			Status:  metav1.ConditionTrue,
+		},
+		DisabledStatus: {
+			Type:    "Disabled",
+			Reason:  "Disabled",
+			Message: "enableMetrics is set to False",
+			Status:  metav1.ConditionTrue,
+		},
+		DegradedStatus: {
+			Type:    "Degraded",
+			Reason:  "Degraded",
+			Message: "Metrics collector deployment not successful",
+			Status:  metav1.ConditionTrue,
+		},
+		NotSupportedStatus: {
+			Type:    "NotSupported",
+			Reason:  "NotSupported",
+			Message: "No Prometheus service found in this cluster",
+			Status:  metav1.ConditionTrue,
+		},
 	}
 )
 
-func ReportStatus(ctx context.Context, client client.Client, i *oav1beta1.ObservabilityAddon, t string, reportStatus bool) {
-	if !reportStatus {
-		return
-	}
-	i.Status.Conditions = []oav1beta1.StatusCondition{
-		{
-			Type:               conditions[t]["type"],
-			Status:             metav1.ConditionTrue,
-			LastTransitionTime: metav1.NewTime(time.Now()),
-			Reason:             conditions[t]["reason"],
-			Message:            conditions[t]["message"],
-		},
+func ReportStatus(ctx context.Context, client client.Client, condition StatusConditionName, addonName, addonNs string) error {
+	newCondition := conditions[condition].DeepCopy()
+	newCondition.LastTransitionTime = metav1.NewTime(time.Now())
+
+	// Fetch the ObservabilityAddon instance in local cluster, and update the status
+	// Retry on conflict
+	obsAddon := &oav1beta1.ObservabilityAddon{}
+	retryErr := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		if err := client.Get(ctx, types.NamespacedName{Name: addonName, Namespace: addonNs}, obsAddon); err != nil {
+			return err
+		}
+
+		if !shouldAppendCondition(obsAddon.Status.Conditions, newCondition) {
+			return nil
+		}
+
+		obsAddon.Status.Conditions = append(obsAddon.Status.Conditions, *newCondition)
+
+		if len(obsAddon.Status.Conditions) > MaxStatusConditionsCount {
```

Comment:
Is this just for convenience, or why do we want only 10?

Reply:
It's an arbitrary number to limit the list size... You think we need more than 10? Previously we were overriding the whole list with the last element, so 10 is better than before already.

Reply:
I mostly just wondered if there was a specific reason for such a limit. Couldn't we just keep adding? Not sure if there is some best practice on this area.

Anything, this should be fine for now.

Reply:
I want to avoid an ever increasing list... And potential associated bugs... Maybe it's over engineered